### PR TITLE
apiextensions: optimize customresource discovery using atomic.Value

### PR DESCRIPTION
**Pull request by nikhita**
Fixes TODO: optimize customresource discovery using `atomic.Value` since writing is infrequent.

**Release note**:

```release-note
NONE
```

/cc sttts liggitt hzxuzhonghu

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go [hunk 1]
```diff
@@ -50,31 +52,44 @@ func (r *versionDiscoveryHandler) ServeHTTP(w http.ResponseWriter, req *http.Req
...
 func (r *versionDiscoveryHandler) unsetDiscovery(gv schema.GroupVersion) {
 	r.discoveryLock.Lock()
 	defer r.discoveryLock.Unlock()
 
-	delete(r.discovery, gv)
+	discoveryMap := r.discoveryStorage.Load().(versionDiscoveryMap)
+	discoveryMap2 := make(versionDiscoveryMap)
+	for k, v := range discoveryMap {
+		discoveryMap2[k] = v
+	}
```
Comment (hzxuzhonghu) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go]: why not put this into a function?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go [hunk 2]
```diff
@@ -50,31 +52,44 @@ func (r *versionDiscoveryHandler) ServeHTTP(w http.ResponseWriter, req *http.Req
 }
 
 func (r *versionDiscoveryHandler) getDiscovery(gv schema.GroupVersion) (*discovery.APIVersionHandler, bool) {
-	r.discoveryLock.RLock()
-	defer r.discoveryLock.RUnlock()
-
-	ret, ok := r.discovery[gv]
+	discoveryMap := r.discoveryStorage.Load().(versionDiscoveryMap)
+	ret, ok := discoveryMap[gv]
```
Comment (hzxuzhonghu) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go]: this may data race

Comment (hzxuzhonghu) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go]: ignore this comment

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go (hunk 1 above)
Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go]: Done. ptal.

**Comment by nikhita** [comment]
@sttts @liggitt ptal, thanks.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go [hunk 3]
```diff
@@ -172,13 +172,17 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
 	}
 
 	versionDiscoveryHandler := &versionDiscoveryHandler{
-		discovery: map[schema.GroupVersion]*discovery.APIVersionHandler{},
-		delegate:  delegateHandler,
+		discoveryStorage: atomic.Value{},
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go]: isn't `atomic.Value{}` the default null value? Then we don't need this initializer.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go [hunk 4]
```diff
@@ -20,15 +20,17 @@ import (
...
 	"k8s.io/apiserver/pkg/endpoints/discovery"
 )
 
+type versionDiscoveryMap map[schema.GroupVersion]*discovery.APIVersionHandler
+
 type versionDiscoveryHandler struct {
-	// TODO, writing is infrequent, optimize this
-	discoveryLock sync.RWMutex
-	discovery     map[schema.GroupVersion]*discovery.APIVersionHandler
+	discoveryLock    sync.RWMutex
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go]: Is that lock still needed?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go (hunk 3 above)
Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go]: Yes, removed.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go (hunk 4 above)
Comment (nikhita) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go]: Yes, the lock is used for the writers (but not the readers).

https://golang.org/pkg/sync/atomic/#Value in the "Example (ReadMostly)"

**Comment by nikhita** [comment]
@sttts @liggitt thoughts on this? If the current behaviour feels good enough, I can close this.

**Comment by liggitt** [comment]
I think the RLock is probably fine

**Comment by nikhita** [comment]
> I think the RLock is probably fine

/close

**Pull request closed without merging**